docs/design/sql/10-distributed-scan.md

```diff
@@ -0,0 +1,164 @@
+
+There are two ways to scan the `IMap` - iterate over the record store directly, or use one of the secondary indexes. During the
+planning stage the proper access method is chosen. 
+
+If there are no indexes on the table, the direct scan is chosen and no further optimization is performed.    
+
+If there are indexes on the table, we analyze the predicate stored in the table of the `TableScan` operator. The entry point
+is `IndexResolver.createIndexScans`. 
+
+First, we split the predicate into conjunctive normal form (CNF). For example, the predicate `a=1 AND b=2` is split into 
+`a=1` and `b=2`, while the predicate `a=1 OR b=2` remains unchanged.
+
+Second, for every sub-predicate we find those that could be used by some index. Assume that `col` is a simple column
+expression, and `exp` is an expression that has only constants or parameters at leaves (i.e. it doesn't refer to other columns). 
+We use the following rules:
+1. `col = exp` can be used with `SORTED` and `HASH` indexes
+1. `col [comparison] exp` can be used with `SORTED` indexes
+1. `col IS NULL/TRUE/FALSE` can be treated as equality expression (with slightly different semantics for `NULL` values)
+1. `col = exp1 OR col = exp2` can be treated as union of two equality predicates 
+
+The result of this step is a map from column to candidate expressions that could be used with indexes. For example, 
+`a=1 AND b>2 AND b<4` is returned as: 
+```
+a -> [=1] 
+b -> [>2], [<4]
+```
+
+Next, we iterate over every index, and try to bind candidates to the index based on the index columns and index type.
+General rules are:
+1. `SORTED` index may use equality and comparison conditions, while `HASH` index may use only equality conditions
+1. Only a continuous prefix of index columns can be bound. E.g. for the index `(a, b, c)` and the condition `a=1 AND c=2` we 
+can bind only `a=1`, while for `a=1 AND b=2` we may bind both `a=1` and `b=2`
+1. All expressions in the prefix except for the last one must be equality conditions. E.g. for the index `(a, b)` and the 
+condition `a>1 AND b>2`, we can use only `a>1`   
+
+The result of this step is a map from index to the filter that should be used. For example, for the expression 
+`a=1 AND b>2 AND b<4`, and indexes on `(a, b)`, `(a, c)` and `(b)`, the result would be:
+```
+index(a, b) -> [a=1, b>2 AND b<4]
+index(a, c) -> [a=1, NULL]
+index(b) -> [b>2 AND b<4]
+```
+
+Once the index filter is built, we calculate the remainder filter, such that `indexFilter AND remainderFilter` is equivalent
+to the original filter. 
+
+For every proposed index, we create a `MapIndexScanPhysicalRel` operator that is added to the planner search space. At the 
+moment we add all viable indexes to the search space. This might become a problem in the future releases, when we have 
+joins and multiple table, because there will be too many alternatives to consider. The solution could be not to add certain
+indexes to the search space based on some heuristics. 
```

Comment:
```suggestion
For every proposed index, we create a `MapIndexScanPhysicalRel` operator that is added to the planner search space. 
Currently, we add all usable indexes to the search space. This might become a problem in future releases, when we have 
joins and multiple tables, because there will be too many alternatives to consider. The solution could be to not add certain
indexes to the search space based on some heuristics. 
```